In [2]:
import numpy as np

### Part 1. Array operations

In [3]:
a = np.array([3,6,9])
b = np.array([12,15,18])

result1 = a + b
print(f'Сумма: {result1}')
result2 = b-a
print(f'Разность: {result2}')
result3 = a*b
print(f'Произведение: {result3}')
result4 = a/b
print(f'Частное: {result4}')
result5 = a*2
print(f'Умножение на число: {result5}')

Сумма: [15 21 27]
Разность: [9 9 9]
Произведение: [ 36  90 162]
Частное: [0.25 0.4  0.5 ]
Умножение на число: [ 6 12 18]


### Part 2. Universal functions

In [9]:
arr = np.array([[1, -2, 3],
                [-4, 5, -6],
                [7, -8, 9]])
abs_result = np.abs(arr)
print(f'abs(arr):\n {abs_result}\n')
sqrt_result = np.sqrt(arr)
print(f'sqrt(arr):\n {sqrt_result}\n')
exp_result = np.exp(arr)
print(f'exp(arr):\n {exp_result}\n')
log_result = np.log(arr)
print(f'log(arr):\n {log_result}\n')
modf_result = np.modf(sqrt_result)
print(f'modf(sqrt_result):\n {modf_result[0]}\n {modf_result[1]}\n')
isnan_result = np.isnan(sqrt_result)
print(f'isnan(sqrt_result):\n {isnan_result}\n')
cos_result = np.cos(arr)
print(f'cos(arr):\n {cos_result}\n')
sin_result = np.sin(arr)
print(f'sin(arr):\n {sin_result}\n')
tan_result = np.tan(arr)
print(f'tan(arr):\n {tan_result}\n')
arccos_result = np.arccos(cos_result)
print(f'arccos(arr):\n {arccos_result}\n')
arcsin_result = np.arcsin(sin_result)
print(f'arcsin(arr):\n {arcsin_result}\n')
arctan_result = np.arctan(tan_result)
print(f'arctan(arr):\n {arctan_result}\n')

abs(arr):
 [[1 2 3]
 [4 5 6]
 [7 8 9]]

sqrt(arr):
 [[1.                nan 1.73205081]
 [       nan 2.23606798        nan]
 [2.64575131        nan 3.        ]]

exp(arr):
 [[2.71828183e+00 1.35335283e-01 2.00855369e+01]
 [1.83156389e-02 1.48413159e+02 2.47875218e-03]
 [1.09663316e+03 3.35462628e-04 8.10308393e+03]]

log(arr):
 [[0.                nan 1.09861229]
 [       nan 1.60943791        nan]
 [1.94591015        nan 2.19722458]]

modf(sqrt_result):
 [[0.                nan 0.73205081]
 [       nan 0.23606798        nan]
 [0.64575131        nan 0.        ]]
 [[ 1. nan  1.]
 [nan  2. nan]
 [ 2. nan  3.]]

isnan(sqrt_result):
 [[False  True False]
 [ True False  True]
 [False  True False]]

cos(arr):
 [[ 0.54030231 -0.41614684 -0.9899925 ]
 [-0.65364362  0.28366219  0.96017029]
 [ 0.75390225 -0.14550003 -0.91113026]]

sin(arr):
 [[ 0.84147098 -0.90929743  0.14112001]
 [ 0.7568025  -0.95892427  0.2794155 ]
 [ 0.6569866  -0.98935825  0.41211849]]

tan(arr):
 [[ 1.55740772  2.18503986 

/home/alexey300479/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: RuntimeWarning: invalid value encountered in sqrt
  
/home/alexey300479/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:10: RuntimeWarning: invalid value encountered in log
  # Remove the CWD from sys.path while we load stuff.


### Part 3. Array shape change

In [11]:
my_array = np.array([[1,2,3,4,5], [6,7,8,9,10]])
display(my_array)
my_array_T = my_array.T
display(my_array_T)

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10]])

array([[ 1,  6],
       [ 2,  7],
       [ 3,  8],
       [ 4,  9],
       [ 5, 10]])

In [12]:
my_array = np.random.randint(0, 10, 20)
my_array.reshape((4,5))

array([[7, 5, 7, 1, 4],
       [7, 5, 5, 9, 1],
       [3, 4, 4, 8, 2],
       [0, 2, 6, 0, 3]])

In [13]:
my_array = np.array([[1,2,3], [11,22,33], [111,222,333]])
my_array.flatten()

array([  1,   2,   3,  11,  22,  33, 111, 222, 333])

### Part 4. Comparison and masks

In [15]:
my_array = np.random.randint(0, 10, (3,4))
my_array

array([[1, 0, 0, 9],
       [5, 4, 7, 6],
       [9, 7, 6, 1]])

In [17]:
my_array < 5

array([[ True,  True,  True, False],
       [False,  True, False, False],
       [False, False, False,  True]])

In [18]:
mask = np.array([1, 0, 1, 0], dtype=bool)
my_array[:, mask]

array([[1, 0],
       [5, 7],
       [9, 6]])

### Part 5. 2D arrays sort

In [20]:
my_array = np.random.randint(0, 10, (4, 6))

In [23]:
my_array

array([[2, 5, 0, 5, 7, 6],
       [4, 0, 5, 0, 6, 1],
       [6, 4, 5, 1, 1, 2],
       [8, 3, 8, 7, 8, 9]])

In [21]:
np.sort(my_array, axis=1)

array([[0, 2, 5, 5, 6, 7],
       [0, 0, 1, 4, 5, 6],
       [1, 1, 2, 4, 5, 6],
       [3, 7, 8, 8, 8, 9]])

In [22]:
np.sort(my_array, axis=0)

array([[2, 0, 0, 0, 1, 1],
       [4, 3, 5, 1, 6, 2],
       [6, 4, 5, 5, 7, 6],
       [8, 5, 8, 7, 8, 9]])

### Part 6. Tasks

#### Common code

In [24]:
first = [x**(1/2) for x in range(100)]
second = [x**(1/3) for x in range(100, 200)]
third = [x/y for x in range(200,300,2) for y in [3,5]]

great_secret = np.array([first, second, third]).T

#### Task 1.
Сколько столбцов содержит массив great_secret?

#### Solution:

In [25]:
great_secret.shape[1]

3

#### Conclusion: 3 is the right answer!

#### Task 2.
Чему равна сумма косинусов элементов первой строки массива great_secret? Ответ округлите до двух знаков после запятой.

#### Solution:

In [30]:
first_row_cos_sum = np.cos(great_secret[0]).sum()
print(f'Сумма косинусов элементов первой строки массива great_secret: {first_row_cos_sum:.2f}')

Сумма косинусов элементов первой строки массива great_secret: 0.16


#### Conclusion: 0.16 is the right answer!!!

#### Task 3.
Чему равна сумма элементов массива great_secret, значение которых больше 50?

#### Solution:

In [31]:
great_secret[great_secret > 50].sum()

5470.0

#### Conclusion: 5470 is the right answer!!!

#### Task 4.
Переведите массив great_secret в одномерную форму. Какое значение в получившемся массиве имеет элемент с индексом 150? Скопируйте ответ из Jupyter Notebook без изменений.

#### Solution:

In [32]:
great_secret.flatten()[150]

7.0710678118654755

#### Conclusion: 7.0710678118654755 is the right answer!

#### Task 5.
Отсортируйте значения столбцов массива great_secret по возрастанию. Чему равна сумма элементов последней строки отсортированного массива? Ответ округлите до двух цифр после запятой.

#### Solution:

In [35]:
last_row_sum = np.sort(great_secret, axis=0)[-1].sum()
print(f'Cумма элементов последней строки отсортированного массива: {last_row_sum:.2f}')

Cумма элементов последней строки отсортированного массива: 115.12


#### Conclusion: 115.12 is the right asnwer!!!